### PR TITLE
Generic Tensor Reduction for Double (SWDEV-284915)

### PR DESCRIPTION
**Pull request by qianfengz**
Adds `miopenDouble` to the public API.

Adds support for miopenReduceTensor() to be used on Tensors of double type data. All the 8 Reduction Operations which are supported with fp32/fp16 Tensors can be used with double Tensors. The alpha/beta scaling factors follow the same specification as what is used by cuddnReduceTensor(),  that is,  for fp32/fp16, alpha/beta should be float type passed by the application; for double, alpha/beta should be double type passed by the application.

This is expected to resolve https://ontrack-internal.amd.com/browse/SWDEV-284915.

<details><summary><b><i>The commits have gone through the following tests:</i></b></summary>

Regular MIOpen tests:
```bash
$ ./bin/test_reduce_test --double --all
$ ./bin/test_reduce_test --all
$ ./bin/test_reduce_test --half --all
```
Manual testing:
```bash
#!/bin/bash

### for fp64
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0 -O 0 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1 -O 0 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2 -O 0 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2,3 -O 0 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0 -O 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1 -O 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2 -O 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2,3 -O 1 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0 -O 2 -N 0 -I 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0 -O 3 -N 0 -I 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0 -O 4 -N 0 -I 1 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1 -O 2 -N 0 -I 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1 -O 3 -N 0 -I 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1 -O 4 -N 0 -I 1 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2 -O 2 -N 0 -I 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2 -O 3 -N 0 -I 1 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2 -O 4 -N 0 -I 1 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2,3 -O 2 -N 0 -I 1 

## skip this on ROCm 3.7 due to unexpected numeric error
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2,3 -O 3 -N 0 -I 1 
## skip this on ROCm 3.7 due to unexpected numeric error
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2,3 -O 4 -N 0 -I 1 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0 -O 5 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0 -O 6 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0 -O 7 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1 -O 5 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1 -O 6 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1 -O 7 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2 -O 5 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2 -O 6 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2 -O 7 

bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2,3 -O 5 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2,3 -O 6 
bin/MIOpenDriver reducefp64 -D 64,3,280,81  -R 0,1,2,3 -O 7 

### for fp32
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 0
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1 -O 0 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2 -O 0 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2,3 -O 0 

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 1 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1 -O 1 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2 -O 1 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2,3 -O 1

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 4 -N 0 -I 1

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1 -O 4 -N 0 -I 1

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2 -O 4 -N 0 -I 1

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2,3 -O 2 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2,3 -O 3 -N 0 -I 1
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2,3 -O 4 -N 0 -I 1

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 5 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 6
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0 -O 7 

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1 -O 5 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1 -O 6 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1 -O 7 

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2 -O 5 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2 -O 6 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2 -O 7 

bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2,3 -O 5
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2,3 -O 6 
bin/MIOpenDriver reduce -D 64,3,280,81  -R 0,1,2,3 -O 7 

### for fp16
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0 -O 0
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1 -O 0 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2 -O 0 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2,3 -O 0 

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0 -O 1 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1 -O 1 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2 -O 1 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2,3 -O 1

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0 -O 2 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0 -O 3 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0 -O 4 -N 0 -I 1

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1 -O 2 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1 -O 3 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1 -O 4 -N 0 -I 1

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2 -O 2 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2 -O 3 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2 -O 4 -N 0 -I 1

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2,3 -O 2 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2,3 -O 3 -N 0 -I 1
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2,3 -O 4 -N 0 -I 1

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0 -O 5 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0 -O 6
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0 -O 7 

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1 -O 5 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1 -O 6 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1 -O 7 

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2 -O 5 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2 -O 6 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2 -O 7 

bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2,3 -O 5
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2,3 -O 6 
bin/MIOpenDriver reducefp16 -D 4,3,60,50  -R 0,1,2,3 -O 7 
```

### CODE REVIEW

**Comment by atamazov** [comment]
@qianfengz Please fix tidy issues. Do not hesitate to ask for help, if needed. Thanks.

**Comment by qianfengz** [comment]
> @qianfengz Please fix tidy issues. Do not hesitate to ask for help, if needed. Thanks.

I am trying to create a docker to be used locally for doing "make analyze".   But it seems from my location,  the libboost-1.72 could not be accessed in the scripts used by MIOpen/Dockerfile.   Any existing docker image can be pulled to do "make analyze" ?

**Review by junliume — COMMENTED** [review]
> libboost-1.72
Could you test if the following URL is not accessible? 
https://boostorg.jfrog.io/artifactory/main/release/1.72.0/source/boost_1_72_0.tar.gz

> > @qianfengz Please fix tidy issues. Do not hesitate to ask for help, if needed. Thanks.
> 
> I am trying to create a docker to be used locally for doing "make analyze". But it seems from my location, the libboost-1.72 could not be accessed in the scripts used by MIOpen/Dockerfile. Any existing docker image can be pulled to do "make analyze" ?

**Comment by atamazov** [comment]
> > @qianfengz Please fix tidy issues. Do not hesitate to ask for help, if needed. Thanks.
> 
> I am trying to create a docker to be used locally for doing "make analyze". But it seems from my location, the libboost-1.72 could not be accessed in the scripts used by MIOpen/Dockerfile. 

> Any existing docker image can be pulled to do "make analyze"?

Make analyze consists of `make cppcheck -j $(nproc)` and `make tidy -j $(nproc)`.

For the former, you need to install cppcheck (see ./dev-requirements.txt):
```
git clone --single-branch https://github.com/danmar/cppcheck
cd cppcheck
git checkout dd05839a7e63ef04afd34711cb3e1e0ef742882f
mkdir build
cd build
cmake ..
make -j $(nproc)
make test
sudo make install
```

**Comment by atamazov** [comment]
@qianfengz For `make tidy`, you can `docker pull rocm/miopen:miopen-rocm-37`. It has everything required for building & running MIOpen _except_ cppcheck. Dependencies reside in `/root/driver/MLOpen/deps_hip` and in `/root/driver/MLOpen/deps_opencl`.

You can also install cppcheck within the container and then `docker commit` it and use your personalized container for both checks.

File: test/reduce_test.cpp [hunk 1]
```diff
@@ -344,6 +347,22 @@ struct verify_reduce_with_indices
...
+        {
+            *reinterpret_cast<double*>(alphaPara) = static_cast<double>(alpha);
+            *reinterpret_cast<double*>(betaPara)  = static_cast<double>(beta);
+        }
+        else
+        {
+            *reinterpret_cast<float*>(alphaPara) = alpha;
+            *reinterpret_cast<float*>(betaPara)  = beta;
+        };
+
```
Comment (atamazov) [inline test/reduce_test.cpp:365]:
```suggestion
        const double alpha64 = alpha;
        const double beta64 = beta;
        const void* const alphaPtr = (std::is_same<T, double>::value) ? &alpha64 : &alpha;
        const void* const betaPtr = (std::is_same<T, double>::value) ? &beta64 : &beta;
```
Then pass alphaPtr and betaPtr.

If compiler or tidy will ask for casting (basically it shouldn't), then use static cast:
```C++
        const void* const alphaPtr = (std::is_same<T, double>::value) ? static_cast<const void*>(&alpha64)...
```

**Review by atamazov — CHANGES_REQUESTED** [review]

File: test/reduce_test.cpp [hunk 2]
```diff
@@ -353,10 +372,10 @@ struct verify_reduce_with_indices
                                 indices_sizeInBytes,
                                 workspace_dev.get(),
                                 ws_sizeInBytes,
-                                static_cast<const void*>(&alpha),
+                                const_cast<const void*>(alphaPara),
```
[the remark this answers is not in the document]
Comment (atamazov) [reply test/reduce_test.cpp:375]: +1 to @junliume , pls see suggestion below.

**Comment by qianfengz** [comment]
> > libboost-1.72
> 
> @qianfengz
> Could you test if the following URL is not accessible?
> https://boostorg.jfrog.io/artifactory/main/release/1.72.0/source/boost_1_72_0.tar.gz

I can manually download boost 1.72 and install it for building MIOpen.   But for "make analyze", the issue seems be more complicated since in Dockerfile,  the breaking of boost downloading also breaks other downloading/installation

**Comment by qianfengz** [comment]
> @qianfengz For `make tidy`, you can `docker pull rocm/miopen:miopen-rocm-37`. It has everything required for building & running MIOpen (including boost) _except_ cppcheck. Dependencies reside in `/root/driver/MLOpen/deps_hip` and in `/root/driver/MLOpen/deps_opencl`. There, build MIOpen as usual and then `make tidy -j $(nproc)`.
> 
> You can also install cppcheck within the container and then `docker commit` it and use your personalized container for both checks.

Thank Artem.  `rocm/miopen:miopen-rocm-37` works perfectly for me.

File: test/reduce_test.cpp (hunk 1 above)
Comment (qianfengz) [reply test/reduce_test.cpp]: Yes.

File: test/reduce_test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (qianfengz) [reply test/reduce_test.cpp]: I guess the compiler always wants us to use  `const void *ptr` in way of `const void *ptr = <xxx>` rather than doing cast at locations where it is used.

**Comment by qianfengz** [comment]
One strange issue is that on  `rocm/miopen:miopen-rocm-37`,  the follow testings

-  `bin/MIOpenDriver reducefp64 -D 64,3,280,81 -R 0,1,2,3 -O 4 -N 0 -I 1` 
- `bin/MIOpenDriver reducefp64 -D 64,3,280,81 -R 0,1,2,3 -O 3 -N 0 -I 1` 
- `bin/test_reduce_test --double --D 64 3 280 81 --R 0 1 2 3 --ReduceOp 4 --N 0 --I 1`
- `bin/test_reduce_test --double --D 64 3 280 81 --R 0 1 2 3 --ReduceOp 3 --N 0 --I 1`

failed to get correct maximum value on my MI100/MI25.  While they are able to pass on  ROCM `3.8`,  `3.10` and `4.1.1` on my M100/MI25 (installed from official repository by me). 

Watching the dumped results,  it can be seen that in cases of failure,  the second largest value rather than the largest one is picked by the GPU kernel,  for example the second largest value is 0.999997, the largest value is 0.999999 or 1.0 by different runnings.   

I am doubting there is compiler issues which cause expression  `0.999997 < 0.999999` return false result.

**Comment by junliume** [comment]
> One strange issue is that on `rocm/miopen:miopen-rocm-37`, the follow testings
> 
> * `bin/MIOpenDriver reducefp64 -D 64,3,280,81 -R 0,1,2,3 -O 4 -N 0 -I 1`
> * `bin/MIOpenDriver reducefp64 -D 64,3,280,81 -R 0,1,2,3 -O 3 -N 0 -I 1`
> * `bin/test_reduce_test --double --D 64 3 280 81 --R 0 1 2 3 --ReduceOp 4 --N 0 --I 1`
> * `bin/test_reduce_test --double --D 64 3 280 81 --R 0 1 2 3 --ReduceOp 3 --N 0 --I 1`
> 
> failed to get correct maximum value on my MI100/MI25. While they are able to pass on ROCM `3.8`, `3.10` and `4.1.1` on my M100/MI25 (installed from official repository by me).
> 
> Watching the dumped results, it can be seen that in cases of failure, the second largest value rather than the largest one is picked by the GPU kernel, for example the second largest value is 0.999997, the largest value is 0.999999 or 1.0 by different runnings.
> 
> I am doubting there are compiler issues which cause expression `0.999997 < 0.999999` return false result.

So the new feature revealed an issue only found in previous version of ROCm? I think all CI nodes have already updated to 4.1 at least, and I doubt if they will patch issues for earlier version of compiler, especially if it is already "fixed" in later versions. :)

**Comment by atamazov** [comment]
@junliume > I think all CI nodes have already updated to 4.1

No, our CI nodes are running 3.7. @shurale-nkn is working on the dockerfile upgrade to 4.2.

**Comment by atamazov** [comment]
@qianfengz
> I am doubting there are compiler issues which cause expression 0.999997 < 0.999999 return false result.

Can you please check with rocm 4.2 release? Please see confluence where to pull it from.

**Comment by atamazov** [comment]
@qianfengz If everything works fine with 4.2, then we can disable failing test cases (workaround) for 3.7 and forget about this problem.

**Comment by qianfengz** [comment]
> @qianfengz If everything works fine with 4.2, then we can disable failing test cases (workaround) for 3.7 and forget about this problem.

Wait some time,  I need download 4.2 from home, could not access repo.radeon.com from the office.

File: src/gemm_v2.cpp [hunk 3]
```diff
@@ -297,6 +297,7 @@ miopenStatus_t CallGemmMIOpenTensile(const Handle& handle,
         ptrA            = Data_t(reinterpret_cast<const int8_t*>(A) + a_offset);
         ptrB            = Data_t(reinterpret_cast<const int8_t*>(B) + b_offset);
         ptrC            = Data_t(reinterpret_cast<int32_t*>(C) + c_offset);
+        break;
```
Comment (atamazov) [inline src/gemm_v2.cpp:300]: :anchor:

**Comment by qianfengz** [comment]
> @qianfengz
> 
> > Wait some time, I need download 4.2 from home, could not access repo.radeon.com from the office.
> 
> You can pull docker image of 4.2 from `compute-artifactory.amd.com:5000/rocm-plus-docker/compute-rocm-rel-4.2:21-STG2`.

Thanks for this information.

**Comment by qianfengz** [comment]
> > @qianfengz
> > > Wait some time, I need download 4.2 from home, could not access repo.radeon.com from the office.
> > 
> > 
> > You can pull docker image of 4.2 from `compute-artifactory.amd.com:5000/rocm-plus-docker/compute-rocm-rel-4.2:21-STG2`.
> 
> Thanks for this information.

No issue found on ROCM 4.2

**Review by junliume — APPROVED** [review]

File: test/reduce_test.cpp [hunk 4]
```diff
@@ -810,6 +881,9 @@ struct reduce_driver : test_driver
                 this->tolerance = 80 * 10;
         };
 
+        if(std::is_same<T, half_float::half>::value)
+            this->tolerance *= this->tolerance * 10.0;
```
Comment (atamazov) [inline test/reduce_test.cpp:885]: This would lead to the constant growth of `tolerance` when called multiple times from `test_drive_impl_1()` via `base_run()`, isn't it?
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/7c34cdeca01e6e4c33b2fb72121aa81a7443fc4e/test/driver.hpp#L1239

**Review by atamazov — COMMENTED** [review]
(Intermediate)

File: test/reduce_test.cpp (hunk 4 above)
Comment (qianfengz) [reply test/reduce_test.cpp:885]: > This would lead to the constant growth of `tolerance` when called multiple times from `test_drive_impl_1()` via `base_run()`, isn't it?
> https://github.com/ROCmSoftwarePlatform/MIOpen/blob/7c34cdeca01e6e4c33b2fb72121aa81a7443fc4e/test/driver.hpp#L1239

Yes, this needs be corrected.

File: test/reduce_test.cpp [hunk 5]
```diff
@@ -589,7 +618,22 @@ struct verify_reduce_no_indices
         return (res);
     }
 
-    tensor<T> gpu() const
+    tensor<float> gpu() const
```
Comment (atamazov) [inline test/reduce_test.cpp:621]: [Q] Why can't we use `T`? Are you sure that it is Ok to use `float` tensor for verification of `double` reduce results?

**Review by atamazov — CHANGES_REQUESTED** [review]
Almost good!

File: test/reduce_test.cpp (hunk 5 above)
Comment (qianfengz) [reply test/reduce_test.cpp:621]: The reason is that in test/driver.cpp,  the verify() function use 
`double threshold = std::numeric_limits<value_type>::epsilon() * tolerance;`    to calculate the threshold,  in which the epsilon() depends on the data type,  this make our codes very difficult to use threshold since the epsilons for half, float and double are very different

File: src/include/miopen/datatype.hpp [hunk 6]
```diff
@@ -127,6 +132,7 @@ inline std::string GetDataTypeKernelParams(miopenDataType_t type)
     ss << " -DMIOPEN_USE_BFP16=" << use_bfp16;
     ss << " -DMIOPEN_USE_INT32=" << use_int32;
     ss << " -DMIOPEN_USE_RNE_BFLOAT16=" << use_rne_bfloat16;
+    ss << " -DMIOPEN_USE_FP64=" << use_fp64;
```
[the remark this answers is not in the document]
Comment (qianfengz) [reply src/include/miopen/datatype.hpp:135]: Understand!  So this a potential point where introduction of fp64 could bring risk to current MIOpen implementation. For those codes other than Generic Reduction, I just made changes in order to remove the compiler complaints.  Can you look into all of those for possible risks ?   Thanks

File: test/driver.hpp [hunk 7]
```diff
@@ -1169,7 +1175,10 @@ void test_drive_impl_1(std::string program_name, std::vector<std::string> as)
     }
     else if(arg_map.count("--double") > 0)
     {
-        throw std::runtime_error("Double is not supported");
+        if(program_name.find("reduce") == std::string::npos)
```
[the remark this answers is not in the document]
Comment (qianfengz) [reply test/driver.hpp:1178]: It is not needed for removing compiler complaints,  just put it here for the tester run-time to give early warning for unsupported using of double.  So do you mean we should not use specific module name in this generic codes ?

[the remark this answers is not in the document]
Comment (qianfengz) [reply test/driver.hpp]: Removed the warnings by [aca973](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/934/commits/bd1610bc043cf1482880be18798de7f29faca973)

[the remark this answers is not in the document]
Comment (atamazov) [reply test/driver.hpp]: Thanks!

File: test/reduce_test.cpp (hunk 5 above)
Comment (atamazov) [reply test/reduce_test.cpp:621]: Ok!

**Review by atamazov — DISMISSED** [review]
LGTM!

**Review by shurale-nkn — DISMISSED** [review]
As I can see, a new data type has been added here, which means need to add tests for reduction with double type. But since we don't have testing stage for double on jenkins, need add custom test with reduce_test --all and attach it for full fp32 stages on target GPUs.

**Comment by atamazov** [comment]
> need add custom test with reduce_test --all and attach it for full fp32 stages on target GPUs.

Theoretically, we should add MIOPEN_TEST_DOUBLE stages, but yes, it seems too expensive to add two stages only to test reduction. Let's add custom reduction tests for double to FULL FP32 stages.

**Comment by atamazov** [comment]
@qianfengz Please _merge from develop_ and add tests. Do not hesitate to ask me if you have any questions or need assistance. Thanks.

**Comment by qianfengz** [comment]
> @qianfengz Please _merge from develop_ and add tests. Do not hesitate to ask me if you have any questions or need assistance. Thanks.

So my understand is that since there is no stage/paralleled sub-stage for `double` defined in the Jekinsfile,  the `bin/test_reduce_test --double --all` has no chance to be executed during the CI.  And so you want me to add `bin/test_reduce_test --double --all`  using `add_custom_test` method so that it can be executed when MIOPEN_TEST_FLOAT and MIOPEN_TEST_ALL are enabled by some stages/sub-stages.   Is this right ?

File: test/CMakeLists.txt [hunk 8]
```diff
@@ -217,7 +217,6 @@ set( LONG_TESTS
     pooling3d.cpp
     soft_max.cpp
     lrn_test.cpp
-    reduce_test.cpp
```
Comment (atamazov) [inline test/CMakeLists.txt:220]: Please revert this. It tests reduce in accordance with the datatype of the stage (as per to MIOPEN_TEST_HALF/IN8/BFLOAT16 or default FP32).

File: test/CMakeLists.txt [hunk 9]
```diff
@@ -1053,3 +1052,7 @@ if(MIOPEN_TEST_CONV)
     COMMAND	$<TARGET_FILE:test_conv2d>	--verbose	--input	1	48	7	7	--weights	1	48	5	5	--pads_strides_dilations	0	0	4	4	1	1
 )
 endif()
+
+if(MIOPEN_TEST_FLOAT)
+    add_custom_test(test_reduce_double SKIP_UNLESS_ALL FLOAT_ENABLED COMMAND  $<TARGET_FILE:test_reduce_test> --double --all)
+endif()
```
Comment (atamazov) [inline test/CMakeLists.txt:1058]: Custom tests are enabled for FLOAT  by default -> let's remove.
gfx908 are disabled by default -> let's enable explicitly.
```suggestion

if(MIOPEN_TEST_FLOAT)
    add_custom_test(test_reduce_double SKIP_UNLESS_ALL GFX908_ENABLED
COMMAND $<TARGET_FILE:test_reduce_test> --double --all --verbose
)
endif()
```

**Review by atamazov — CHANGES_REQUESTED** [review]

**Comment by atamazov** [comment]
> ...And so you want me to add `bin/test_reduce_test --double --all` using `add_custom_test` method so that it can be executed when MIOPEN_TEST_FLOAT and MIOPEN_TEST_ALL are enabled by some stages/sub-stages. Is this right ?

Correct!

File: test/CMakeLists.txt (hunk 8 above)
Comment (atamazov) [reply test/CMakeLists.txt:220]: FYI this block of code ensures that reduce_test is used with FLOAT and HALF on gfx906/900 and gfx908:
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/96d7086fa0113e2109de44f870dbae1c340b48fc/test/CMakeLists.txt#L93-L105

Comment (atamazov) [reply test/CMakeLists.txt:220]: > Please revert this. It tests reduce in accordance with the datatype of the stage (as per to MIOPEN_TEST_HALF/IN8/BFLOAT16 or default FP32).

Please ignora the above. Currently LONG_TESTS is not used at all. It's an old leftover that should be eventually removed.

**Review by atamazov — APPROVED** [review]
LGTM!

File: test/CMakeLists.txt [hunk 10]
```diff
@@ -1053,3 +1052,7 @@ if(MIOPEN_TEST_CONV)
     COMMAND	$<TARGET_FILE:test_conv2d>	--verbose	--input	1	48	7	7	--weights	1	48	5	5	--pads_strides_dilations	0	0	4	4	1	1
 )
 endif()
+
+if(MIOPEN_TEST_FLOAT)
```
Comment (atamazov) [inline test/CMakeLists.txt:1056]: [Notice] Well, this `if` is not required actually (but there is no harm). Ok.

**Review by shurale-nkn — APPROVED** [review]

**Comment by atamazov** [comment]
## @qianfengz :tada:

File: src/reducetensor.cpp [hunk 11]
```diff
@@ -605,6 +622,8 @@ void ReduceTensorDescriptor::ReduceTensor(const Handle& handle,
 
         std::string param2 = param + " -DCK_PARAM_GRIDSIZE=" + std::to_string(gridSize_2) + " ";
 
+        std::string network_config2 = network_config + "_C2";
```
Comment (atamazov) [inline src/reducetensor.cpp:625]: @qianfengz Why this change (appending "_C2") is necessary?

File: src/solver/conv_asm_1x1u.cpp [hunk 12]
```diff
@@ -308,6 +308,9 @@ bool PerformanceConfigConvAsm1x1U::IsValid(const ConvolutionContext& config) con
 
 void PerformanceConfigConvAsm1x1U::HeuristicInit(const ConvolutionContext& config)
 {
+    if(config.in_data_type == miopenDouble)
```
Comment (atamazov) [inline src/solver/conv_asm_1x1u.cpp:311]: It seems like this change is not needed. `IsApplicable()` of this Solver should refuse configs with double.

**Review by atamazov — COMMENTED** [review]
Post-merge review 1 (part of work on https://ontrack-internal.amd.com/browse/SWDEV-293780)

**Comment by atamazov** [comment]
It seems like this PR is the reason of https://ontrack-internal.amd.com/browse/SWDEV-293780. I am in the process of double confirming this. Meanwhile please check from your side if this PR contains something wrong.

File: src/reducetensor.cpp (hunk 11 above)
Comment (qianfengz) [reply src/reducetensor.cpp:625]: The second time and the first time use directly kernels since even though they are solving the same issue, so they should have different key for mapping in the kernel cache. 

One other thing that I want to mention is that, kernel cache is actually not queried,  since `Handle::AddKernel() ` actually only queries the Program cache,   and we does not use `Handle::HasKernel() + Handle::GetKernel()` here

Comment (atamazov) [reply src/reducetensor.cpp:625]: @qianfengz 
> The second time and the first time use different kernels even though they are for solving the same problem, so they should have different key for mapping in the kernel cache.

I do not understand, sorry.

Comment (atamazov) [reply src/reducetensor.cpp:625]: ?

File: src/solver/conv_asm_1x1u.cpp (hunk 12 above)
Comment (atamazov) [reply src/solver/conv_asm_1x1u.cpp:311]: ?
